python/ray/dag/dag_node_operation.py

```diff
@@ -174,33 +232,63 @@ def _select_next_nodes(
         A list of _DAGOperationGraphNodes to be placed into the corresponding
         execution schedules.
     """
+    # If a node is a NCCL compute node, it updates the `ready_collective_nodes`
+    # of all the nodes in its collective group. It is removed from the candidate
+    # list if the collective group is not ready.
+    for actor, candidates in actor_to_candidates.items():
+        ready_candidates: List[_DAGOperationGraphNode] = []
+        for node in candidates:
+            if node.is_nccl_compute:
+                for collective_node_metadata in node.collective_group:
+                    task_idx, op_type = (
+                        collective_node_metadata[0],
+                        collective_node_metadata[1],
+                    )
+                    collective_node = graph[task_idx][op_type]
+                    collective_node.ready_collective_nodes.add(
+                        (node.task_idx, node.operation.type)
+                    )
+            if node.is_ready:
```

Comment:
Since we add to ready_collective_nodes while iterating, I think this will only add one ready node per collective group as a candidate, but we want to add all ready nodes per collective group.

Reply:
Yes, only one node will be ready in the group. After that, all the nodes in the group are added in the branch of `elif top_priority_node.is_nccl_collective:`.

The behavior is that when one node is popped from `actor_to_candidates`. It should be ready to be executed. If only partial nodes in the collective group are in `actor_to_candidates`, and one is popped, it would break this behaviour.

So when a node has in-degree as 0, it is added to `actor_to_candidates`. If a node is a NCCL collective node, it updates the `ready_collective_nodes` of all the nodes in its collective group. It is removed from the candidate list if the collective group is not ready.